test/test_dag.py

```diff
@@ -344,944 +334,885 @@ def test_auxiliary_steps(clear_now_and_after):
-        #
-        #  A1 <- B1 <- B2 <- B3
-        #   |
-        #   |
-        #  \/
-        #  A2 <- B2 <- C1 <- C2 <- C4
-        #                    C3
-        #
-        script = SoS_Script(textwrap.dedent('''
-        [A_1]
-        input: 'B1.txt'
-        output: 'A1.txt'
-        run:
-            touch A1.txt
-
-        [A_2]
-        depends:  'B2.txt'
-        output: 'A2.txt'
-        run:
-            touch A2.txt
-
-        [B1: provides='B1.txt']
-        depends: 'B2.txt'
-        run:
-            touch B1.txt
-
-        [B2: provides='B2.txt']
-        depends: 'B3.txt', 'C1.txt'
-        run:
-            touch B2.txt
-
-        [B3: provides='B3.txt']
-        run:
-            touch B3.txt
-
-        [C1: provides='C1.txt']
-        depends: 'C2.txt', 'C3.txt'
-        run:
-            touch C1.txt
-
-        [C2: provides='C2.txt']
-        depends: 'C4.txt'
-        run:
-            touch C2.txt
+def test_cycle():
+    '''Test cycle detection of DAG'''
+    #
+    #  A.txt --> B.txt
+    #
+    #  B.txt --> C.txt
+    #
+    #  C.txt --> A.txt
+    #
+    script = SoS_Script(textwrap.dedent('''
+    [A_1]
+    input: 'A.txt'
+    output: 'B.txt'
 
-        [C3: provides='C3.txt']
-        depends: 'C4.txt'
-        run:
-            touch C3.txt
+    [A_2]
+    output: 'C.txt'
 
-        [C4: provides='C4.txt']
-        run:
-            touch C4.txt
+    [A_3]
+    output: 'A.txt'
+    '''))
+    # the workflow should call step K for step C_2, but not C_3
+    wf = script.workflow()
+    pytest.raises(RuntimeError, Base_Executor(wf).initialize_dag)
 
+def test_long_chain(clear_now_and_after):
+    '''Test long make file style dependencies.'''
+    #
+    for f in [
+            'A1.txt', 'A2.txt', 'C2.txt', 'B2.txt', 'B1.txt', 'B3.txt',
+            'C1.txt', 'C3.txt', 'C4.txt'
+    ]:
+        clear_now_and_after(f)
+    #
+    #  A1 <- B1 <- B2 <- B3
+    #   |
+    #   |
+    #  \/
+    #  A2 <- B2 <- C1 <- C2 <- C4
+    #                    C3
+    #
+    script = SoS_Script(textwrap.dedent('''
+    [A_1]
+    input: 'B1.txt'
+    output: 'A1.txt'
+    run:
+        touch A1.txt
+
+    [A_2]
+    depends:  'B2.txt'
+    output: 'A2.txt'
+    run:
+        touch A2.txt
+
+    [B1: provides='B1.txt']
+    depends: 'B2.txt'
+    run:
+        touch B1.txt
+
+    [B2: provides='B2.txt']
+    depends: 'B3.txt', 'C1.txt'
+    run:
+        touch B2.txt
+
+    [B3: provides='B3.txt']
+    run:
+        touch B3.txt
+
+    [C1: provides='C1.txt']
+    depends: 'C2.txt', 'C3.txt'
+    run:
+        touch C1.txt
+
+    [C2: provides='C2.txt']
+    depends: 'C4.txt'
+    run:
+        touch C2.txt
+
+    [C3: provides='C3.txt']
+    depends: 'C4.txt'
+    run:
+        touch C3.txt
+
+    [C4: provides='C4.txt']
+    run:
+        touch C4.txt
+
+    '''))
+    # the workflow should call step K for step C_2, but not C_3
+    wf = script.workflow()
+    #env.verbosity = 4
+    dag = Base_Executor(wf).initialize_dag()
```

Comment:
We have `get_initial_dag`, right?

Reply:
That makes sense. Originally I did not change the ones with `Base_Executor(wf).run()` used twice, but the second one could be removed since the unlink statements were unnecessary.